pkg/cdi/spec.go

```diff
@@ -262,3 +272,69 @@ func validateSpec(raw *cdi.Spec) error {
 	}
 	return nil
 }
+
+// GenerateSpecName generates a vendor+class scoped Spec file name. The
+// name can be passed to WriteSpec() to write a Spec file to the file
+// system.
+//
+// vendor and class should match the vendor and class of the CDI Spec.
+// The file name is generated without a ".json" or ".yaml" extension.
+// The caller can append the desired extension to choose a particular
+// encoding. Otherwise WriteSpec() will use its default encoding.
+//
+// This function always returns the same name for the same vendor/class
+// combination. Therefore it cannot be used as such to generate multiple
+// Spec file names for a single vendor and class.
+func GenerateSpecName(vendor, class string) string {
+	return vendor + "-" + class
+}
+
+// GenerateTransientSpecName generates a vendor+class scoped transient
+// Spec file name. The name can be passed to WriteSpec() to write a Spec
+// file to the file system.
+//
+// Transient Specs are those whose lifecycle is tied to that of some
+// external entity, for instance a container. vendor and class should
+// match the vendor and class of the CDI Spec. transientID should be
+// unique among all CDI users on the same host that might generate
+// transient Spec files using the same vendor/class combination. If
+// the external entity to which the lifecycle of the tranient Spec
+// is tied to has a unique ID of its own, then this is usually a
+// good choice for transientID.
+//
+// The file name is generated without a ".json" or ".yaml" extension.
+// The caller can append the desired extension to choose a particular
+// encoding. Otherwise WriteSpec() will use its default encoding.
+func GenerateTransientSpecName(vendor, class, transientID string) string {
+	transientID = strings.ReplaceAll(transientID, "/", "_")
+	return vendor + "-" + class + "_" + transientID
+}
+
+// GenerateNameForSpec generates a name for the given Spec using
+// GenerateSpecName with the vendor and class taken from the Spec.
+// On success it returns the generated name and a nil error. If
+// the Spec does not contain a valid vendor or class, it returns
+// an empty name and a non-nil error.
+func GenerateNameForSpec(raw *cdi.Spec) (string, error) {
+	vendor, class := ParseQualifier(raw.Kind)
+	if vendor == "" {
+		return "", errors.Errorf("invalid vendor/class %q in Spec", raw.Kind)
+	}
+
+	return GenerateSpecName(vendor, class), nil
+}
+
+// GenerateNameForTransientSpec generates a name for the given transient
+// Spec using GenerateTransientSpecName with the vendor and class taken
+// from the Spec. On success it returns the generated name and a nil error.
+// If the Spec does not contain a valid vendor or class, it returns an
+// an empty name and a non-nil error.
+func GenerateNameForTransientSpec(raw *cdi.Spec, transientID string) (string, error) {
+	vendor, class := ParseQualifier(raw.Kind)
+	if vendor == "" {
+		return "", errors.Errorf("invalid vendor/class %q in Spec", raw.Kind)
+	}
```

Comment:
Note that the test driver cannot use any of these functions because the kubelet plugin might run inside the e2e.test binary (i.e. outside of the cluster) and just manipulate the nodes from remote (using port forwarding and the equivalent of `kubectl exec dd/rm` to create and remove files) - see https://github.com/pohly/kubernetes/pull/15 and in particular the IO abstraction in [test/e2e/dra/test-driver/app/kubeletplugin.go](https://github.com/pohly/kubernetes/pull/15/commits/dba16e41b7590d886c76a9b113449677c8d3eb85#diff-f20c62b4b40543cf14c979f85bfe98c0ab0e2b1b714458423d64387977df2507).

The API for generating the names might be useful for the test driver if it included the full path, but not as it stands right now because only WriteSpec internally creates the full path.

Reply:
@klihub would it make sense to add an API to get the file path for the written spec? This would allow test driver to get the file path and transfer the file to the remote location.

Reply:
Alternatively Cache.WriteSpec can return path to the written spec. @pohly would it help test driver use case?

Reply:
@bart0sh We can update `WriteSpec()` to return `(path string, err error)` if that helps.

Reply:
@pohly  wdyt?

Reply:
The test driver cannot call `WriteSpec`. That would write the file on the host where the e2e.test runs, not on the host where kubelet and the container runtime run.

Reply:
Can the e2e transfer the file to the host where runtime can access it?

Reply:
Yes. But it doesn't have write access to where this package wants to write the file (`/var/run/cdi`), so writing it first locally and then copying to the remote host won't work.

Reply:
@pohly the code that creates a CDI file is [here](https://github.com/pohly/kubernetes/blob/dynamic-resource-allocation/test/e2e/dra/test-driver/app/kubeletplugin.go#L125:L157) as far as I understand. It still creates a file on the file system. Can you point us out to the  e2e code that uses this file(transfers it to a remote location?)?

Reply:
> It still creates a file on the file system.

No. It encodes the spec in memory (`buffer, err := json.Marshal(*cdiSpec)`) then creates the file on the remote remote host (`ex.fileOps.Create(filePath, buffer)`). The implementation of that file creation is in https://github.com/pohly/kubernetes/blob/c15323637d09a28d7100527a38a2daf3e20e9746/test/e2e/storage/drivers/proxy/io.go#L65-L71

Reply:
@pohly Thanks. Now I more or less understand how it works.

With the proposed API change it will not be possible I guess as newSpec is a private method there.
So, we need to either leave NewSpec public or modify an API that it can be used by e2e.